tests/sycl/relational.cpp

```diff
@@ -0,0 +1,237 @@
+
+  template<typename DT, int D, std::enable_if_t<D<=4, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return std::get<D>(std::make_tuple(
+      v.s0(),
+      vec<DT, 1>(v.s0()),
+      vec<DT, 2>(v.s0(), v.s1()),
+      vec<DT, 3>(v.s0(), v.s1(), v.s2()),
+      vec<DT, 4>(v.s0(), v.s1(), v.s2(), v.s3())));
+  }
+  template<typename DT, int D, std::enable_if_t<D==8, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return vec<DT, 8>(v.s0(), v.s1(), v.s2(), v.s3(), v.s4(), v.s5(), v.s6(), v.s7());
+  }
+  template<typename DT, int D, std::enable_if_t<D==16, int> = 0>
+  auto get_rel_input(cl::sycl::vec<DT, 16> v) {
+    return v;
+  }
+
+  // runtime indexed access to vector elements
+  // this could be a single function with constexpr if in C++17
+  // could also be done by using knowledge of the internal structure,
+  // but I wanted to avoid that.
+  template<typename T, std::enable_if_t<vector_length_v<T> == 0, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx == 0);
+    return v;
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 1, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    return v.x();
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 2, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    if(idx==0) return v.x();
+    return v.y();
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 3, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    if(idx==0) return v.x();
+    if(idx==1) return v.y();
+    return v.z();
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 4, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    if(idx==0) return v.x();
+    if(idx==1) return v.y();
+    if(idx==2) return v.z();
+    return v.w();
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 8, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    if(idx==0) return v.s0();
+    if(idx==1) return v.s1();
+    if(idx==2) return v.s2();
+    if(idx==3) return v.s3();
+    if(idx==4) return v.s4();
+    if(idx==5) return v.s5();
+    if(idx==6) return v.s6();
+    if(idx==7) return v.s7();
+    return v.s7();
+  }
+  template<typename T, std::enable_if_t<vector_length_v<T> == 16, int> = 0>
+  auto comp(T v, size_t idx) {
+    assert(idx < vector_length_v<T>);
+    if(idx==0) return v.s0();
+    if(idx==1) return v.s1();
+    if(idx==2) return v.s2();
+    if(idx==3) return v.s3();
+    if(idx==4) return v.s4();
+    if(idx==5) return v.s5();
+    if(idx==6) return v.s6();
+    if(idx==7) return v.s7();
+    if(idx==8) return v.s8();
+    if(idx==9) return v.s9();
+    if(idx==10) return v.sA();
+    if(idx==11) return v.sB();
+    if(idx==12) return v.sC();
+    if(idx==13) return v.sD();
+    if(idx==14) return v.sE();
+    if(idx==15) return v.sF();
```

Comment:
SYCL 2020 has introduced `vec::operator[]`, so I think you could simplify this quite a bit :-)

Reply:
Done.